File: 0876-middle-of-the-linked-list/python/Solution.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def middleNode(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head:
            return head
        fast = head
        slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        return slow
    def middleNodeTwoPass(self, head):
        if not head:
            return head
        p = head
        length = 0
        while p:
            p = p.next
            length += 1
        p = head
        for i in range(length / 2):
            p = p.next
        return p

File: 0876-middle-of-the-linked-list/java/Solution.java
class Solution {
  public class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
      val = x;
    }
  }
  
  // Time: O(n), Space: O(1) 遍历两次
  public ListNode middleNode(ListNode head) {
    ListNode p = head;
    int len = 0;
    for (; p != null; p = p.next) ++len;
    p = head;
    for (int i = 0; i < len/2; ++i) p = p.next;
    return p;
  }
  // Time: O(n), Space: O(1) 遍历一次
  public ListNode getMiddleNodeOnePass(ListNode head) {
    ListNode fast = head, slow = head;
    while (fast != null && fast.next != null) {
      fast = fast.next.next;
      slow = slow.next;
    }
    return slow;
  }
}